File: core/src/androidTest/java/com/emarsys/core/worker/CoreCompletionHandlerMiddlewareTest.java
package com.emarsys.core.worker;

import android.os.Handler;
import android.os.Looper;
import android.os.Message;

import com.emarsys.core.CoreCompletionHandler;
import com.emarsys.core.concurrency.CoreSdkHandlerProvider;
import com.emarsys.core.database.repository.Repository;
import com.emarsys.core.database.repository.SqlSpecification;
import com.emarsys.core.request.model.CompositeRequestModel;
import com.emarsys.core.request.model.RequestMethod;
import com.emarsys.core.request.model.RequestModel;
import com.emarsys.core.request.model.specification.FilterByRequestId;
import com.emarsys.core.response.ResponseModel;
import com.emarsys.test.util.HandlerUtils;
import com.emarsys.test.util.TimeoutUtils;

import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TestRule;
import org.mockito.ArgumentCaptor;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotNull;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.verifyZeroInteractions;
import static org.mockito.Mockito.when;

public class CoreCompletionHandlerMiddlewareTest {

    private CoreCompletionHandler coreCompletionHandler;
    private Repository<RequestModel, SqlSpecification> requestRepository;
    private Worker worker;

    private String expectedId;
    private CoreCompletionHandlerMiddleware middleware;
    private ArgumentCaptor<Message> captor;
    private Handler uiHandler;
    private Handler coreSdkHandler;

    @Rule
    public TestRule timeout = TimeoutUtils.getTimeoutRule();

    @Before
    @SuppressWarnings("unchecked")
    public void setup() {
        expectedId = "expectedId";

        worker = mock(Worker.class);
        coreCompletionHandler = mock(CoreCompletionHandler.class);
        requestRepository = mock(Repository.class);
        uiHandler = new Handler(Looper.getMainLooper());
        coreSdkHandler = new CoreSdkHandlerProvider().provideHandler();
        middleware = new CoreCompletionHandlerMiddleware(worker, requestRepository, uiHandler, coreSdkHandler, coreCompletionHandler);
        captor = ArgumentCaptor.forClass(Message.class);
    }

    @Test
    public void testConstructor_handlerShouldNotBeNull() {
        assertNotNull(middleware.coreSDKHandler);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testConstructor_workerShouldNotBeNull() {
        new CoreCompletionHandlerMiddleware(null, requestRepository, uiHandler, coreSdkHandler, coreCompletionHandler);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testConstructor_queueShouldNotBeNull() {
        new CoreCompletionHandlerMiddleware(worker, null, uiHandler, coreSdkHandler, coreCompletionHandler);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testConstructor_coreCompletionHandlerShouldNotBeNull() {
        new CoreCompletionHandlerMiddleware(worker, requestRepository, uiHandler, coreSdkHandler, null);
    }


    @Test(expected = IllegalArgumentException.class)
    public void testConstructor_uiHandlerShouldNotBeNull() {
        new CoreCompletionHandlerMiddleware(worker, requestRepository, null, coreSdkHandler, coreCompletionHandler);
    }

    @Test(expected = IllegalArgumentException.class)
    public void testConstructor_coreHandlerShouldNotBeNull() {
        new CoreCompletionHandlerMiddleware(worker, requestRepository, uiHandler, null, coreCompletionHandler);
    }

    @Test
    public void testOnSuccess() {
        ArgumentCaptor<FilterByRequestId> captor = ArgumentCaptor.forClass(FilterByRequestId.class);

        ResponseModel expectedModel = createResponseModel(200);

        middleware.onSuccess(expectedId, expectedModel);

        HandlerUtils.waitForEventLoopToFinish(coreSdkHandler);
        HandlerUtils.waitForEventLoopToFinish(uiHandler);

        verify(worker).unlock();
        verify(worker).run();
        verifyNoMoreInteractions(worker);

        verify(requestRepository).remove(captor.capture());
        FilterByRequestId filter = captor.getValue();
        assertEquals(expectedModel.getRequestModel().getId(), filter.getArgs()[0]);

        verify(coreCompletionHandler).onSuccess(expectedId, expectedModel);
    }

    @Test
    public void testOnSuccess_callsHandlerPost() throws Exception {
        Handler handler = mock(Handler.class);
        middleware.coreSDKHandler = handler;

        middleware.onSuccess(expectedId, createResponseModel(200));

        verify(handler).sendMessageAtTime(captor.capture(), any(Long.class));

        Runnable runnable = captor.getValue().getCallback();
        runnable.run();
        verify(worker).run();
    }

    @Test
    public void testOnSuccess_withCompositeModel() {
        String[] ids = new String[]{"id1", "id2", "id3"};

        ArgumentCaptor<String> captor = ArgumentCaptor.forClass(String.class);

        RequestModel requestModel = new CompositeRequestModel(
                "https://emarsys.com",
                RequestMethod.POST,
                null,
                new HashMap<String, String>(),
                100,
                900000,
                ids);

        ResponseModel responseModel = new ResponseModel.Builder()
                .statusCode(200)
                .message("OK")
                .headers(new HashMap<String, List<String>>())
                .body("{'key': 'value'}")
                .requestModel(requestModel)
                .build();

        middleware.onSuccess("0", responseModel);

        HandlerUtils.waitForEventLoopToFinish(coreSdkHandler);
        HandlerUtils.waitForEventLoopToFinish(uiHandler);

        verify(middleware.coreCompletionHandler, times(3)).onSuccess(captor.capture(), eq(responseModel));
        List<String> capturedIds = captor.getAllValues();
        assertEquals(Arrays.asList(ids), capturedIds);
    }

    @Test
    public void testOnError_4xx() {
        ArgumentCaptor<FilterByRequestId> captor = ArgumentCaptor.forClass(FilterByRequestId.class);

        ResponseModel expectedModel = createResponseModel(403);

        middleware.onError(expectedId, expectedModel);

        HandlerUtils.waitForEventLoopToFinish(coreSdkHandler);
        HandlerUtils.waitForEventLoopToFinish(uiHandler);

        verify(requestRepository).remove(captor.capture());
        FilterByRequestId filter = captor.getValue();
        assertEquals(expectedModel.getRequestModel().getId(), filter.getArgs()[0]);
        verify(coreCompletionHandler).onError(expectedId, expectedModel);

        verify(worker).unlock();
        verify(worker).run();
        verifyNoMoreInteractions(worker);
    }

    @Test
    public void testOnError_4xx_callsHandlerPost() {
        Handler handler = mock(Handler.class);
        middleware.coreSDKHandler = handler;

        middleware.onError(expectedId, createResponseModel(401));

        verify(handler).sendMessageAtTime(captor.capture(), any(Long.class));
        Runnable runnable = captor.getValue().getCallback();
        runnable.run();
        verify(worker).run();
    }

    @Test
    public void testOnError_408_shouldHandleErrorAsRetriable() {
        ResponseModel expectedModel = createResponseModel(408);

        middleware.onError(expectedId, expectedModel);

        HandlerUtils.waitForEventLoopToFinish(coreSdkHandler);
        HandlerUtils.waitForEventLoopToFinish(uiHandler);

        verify(worker).unlock();
        verifyNoMoreInteractions(worker);

        verifyZeroInteractions(coreCompletionHandler);
        verifyZeroInteractions(requestRepository);
    }

    @Test
    public void testOnError_5xx() {
        ResponseModel expectedModel = createResponseModel(500);

        middleware.onError(expectedId, expectedModel);

        HandlerUtils.waitForEventLoopToFinish(coreSdkHandler);
        HandlerUtils.waitForEventLoopToFinish(uiHandler);

        verify(worker).unlock();
        verifyNoMoreInteractions(worker);

        verifyZeroInteractions(coreCompletionHandler);
        verifyZeroInteractions(requestRepository);
    }

    @Test
    public void testOnError_4xx_withCompositeModel() {
        String[] ids = new String[]{"id1", "id2", "id3"};

        ArgumentCaptor<String> captor = ArgumentCaptor.forClass(String.class);

        RequestModel requestModel = new CompositeRequestModel(
                "https://emarsys.com",
                RequestMethod.POST,
                null,
                new HashMap<String, String>(),
                100,
                900000,
                ids);

        ResponseModel responseModel = new ResponseModel.Builder()
                .statusCode(400)
                .message("Bad Request")
                .headers(new HashMap<String, List<String>>())
                .body("{'key': 'value'}")
                .requestModel(requestModel)
                .build();

        middleware.onError("0", responseModel);

        HandlerUtils.waitForEventLoopToFinish(coreSdkHandler);
        HandlerUtils.waitForEventLoopToFinish(uiHandler);

        verify(middleware.coreCompletionHandler, times(3)).onError(captor.capture(), eq(responseModel));
        List<String> capturedIds = captor.getAllValues();
        assertEquals(Arrays.asList(ids), capturedIds);
    }

    @Test
    public void testOnError_withException() {
        Exception expectedException = new Exception("Expected exception");

        middleware.onError(expectedId, expectedException);

        HandlerUtils.waitForEventLoopToFinish(coreSdkHandler);
        HandlerUtils.waitForEventLoopToFinish(uiHandler);

        verify(worker).unlock();
        verifyNoMoreInteractions(worker);

        verify(coreCompletionHandler).onError(expectedId, expectedException);
        verifyNoMoreInteractions(coreCompletionHandler);

        verifyZeroInteractions(requestRepository);
    }

    private ResponseModel createResponseModel(int statusCode) {
        RequestModel requestModel = mock(RequestModel.class);
        when(requestModel.getId()).thenReturn(expectedId);

        return new ResponseModel.Builder()
                .statusCode(statusCode)
                .body("body")
                .message("message")
                .requestModel(requestModel)
                .build();
    }
}

File: core/src/androidTest/java/com/emarsys/core/worker/CoreCompletionHandlerMiddlewareTest.kt
package com.emarsys.core.worker

import android.os.Handler
import android.os.Looper
import android.os.Message
import com.emarsys.core.CoreCompletionHandler
import com.emarsys.core.concurrency.ConcurrentHandlerHolderFactory
import com.emarsys.core.database.repository.Repository
import com.emarsys.core.database.repository.SqlSpecification
import com.emarsys.core.handler.ConcurrentHandlerHolder
import com.emarsys.core.handler.SdkHandler
import com.emarsys.core.request.model.CompositeRequestModel
import com.emarsys.core.request.model.RequestMethod
import com.emarsys.core.request.model.RequestModel
import com.emarsys.core.request.model.specification.FilterByRequestIds
import com.emarsys.core.response.ResponseModel
import com.emarsys.testUtil.mockito.ThreadSpy
import io.kotest.matchers.ints.shouldBeLessThanOrEqual
import io.kotest.matchers.shouldBe
import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.ArgumentCaptor
import org.mockito.kotlin.argumentCaptor
import org.mockito.kotlin.capture
import org.mockito.kotlin.eq
import org.mockito.kotlin.mock
import org.mockito.kotlin.timeout
import org.mockito.kotlin.times
import org.mockito.kotlin.verify
import org.mockito.kotlin.verifyNoInteractions
import org.mockito.kotlin.verifyNoMoreInteractions
import org.mockito.kotlin.whenever
import java.util.concurrent.CountDownLatch

class CoreCompletionHandlerMiddlewareTest {
    private lateinit var coreCompletionHandler: CoreCompletionHandler
    private lateinit var requestRepository: Repository<RequestModel, SqlSpecification>
    private lateinit var mockWorker: Worker
    private lateinit var expectedId: String
    private lateinit var middleware: CoreCompletionHandlerMiddleware
    private lateinit var captor: ArgumentCaptor<Message>
    private lateinit var uiHandler: Handler
    private lateinit var concurrentHandlerHolder: ConcurrentHandlerHolder


    @BeforeEach
    fun setup() {
        expectedId = "expectedId"
        mockWorker = mock()
        coreCompletionHandler = mock()
        requestRepository = mock()
        uiHandler = Handler(Looper.getMainLooper())
        concurrentHandlerHolder = ConcurrentHandlerHolderFactory.create()

        middleware = CoreCompletionHandlerMiddleware(
            mockWorker,
            requestRepository,
            concurrentHandlerHolder,
            coreCompletionHandler
        )
        captor = ArgumentCaptor.forClass(Message::class.java)
    }

    @Test
    fun testOnSuccess() {
        val captor = ArgumentCaptor.forClass(FilterByRequestIds::class.java)
        val expectedModel = createResponseModel(200)

        middleware.onSuccess(expectedId, expectedModel)

        waitForEventLoopToFinish(concurrentHandlerHolder.coreHandler)
        waitForEventLoopToFinish(uiHandler)
        runBlocking {
            verify(mockWorker).unlock()
            verify(mockWorker).run()
            verifyNoMoreInteractions(mockWorker)
            verify(requestRepository).remove(capture<FilterByRequestIds>(captor))
            val filter = captor.value

            filter.selectionArgs[0] shouldBe expectedModel.requestModel.id

            verify(coreCompletionHandler).onSuccess(expectedId, expectedModel)
        }
    }

    @Test
    fun testOnSuccess_withCompositeModel_withManyRequestModels() {
        val ids = Array(2040) { i -> "id$i" }

        val responseModel = createResponseModel(createRequestModel(ids))

        middleware.onSuccess("0", responseModel)

        waitForEventLoopToFinish(concurrentHandlerHolder.coreHandler)
        waitForEventLoopToFinish(uiHandler)

        argumentCaptor<FilterByRequestIds>().apply {
            runBlocking {
                verify(requestRepository, times(41)).remove(capture())
            }
        }

    }

    @Test
    fun testOnSuccess_callsHandlerPost() {
        middleware.concurrentHandlerHolder = concurrentHandlerHolder
        val threadSpy = ThreadSpy<Unit>()
        whenever(mockWorker.unlock()).thenAnswer(threadSpy)
        middleware.onSuccess(expectedId, createResponseModel(200))
        threadSpy.verifyCalledOnCoreSdkThread()
        verify(mockWorker, timeout(50)).run()
    }

    @Test
    fun testOnSuccess_withCompositeModel() {
        val ids = Array(2040) { i -> "id$i" }
        val responseModel = createResponseModel(createRequestModel(ids))

        middleware.onSuccess("0", responseModel)

        waitForEventLoopToFinish(concurrentHandlerHolder.coreHandler)
        waitForEventLoopToFinish(uiHandler)

        argumentCaptor<String>().apply {
            verify(middleware.coreCompletionHandler, times(2040))!!.onSuccess(
                capture(),
                eq(responseModel)
            )
            allValues shouldBe ids.toList()
        }
    }

    @Test
    fun testOnError_4xx() {
        val expectedModel = createResponseModel(403)

        middleware.onError(expectedId, expectedModel)

        waitForEventLoopToFinish(concurrentHandlerHolder.coreHandler)
        waitForEventLoopToFinish(uiHandler)

        argumentCaptor<FilterByRequestIds>().apply {
            runBlocking {
                verify(requestRepository).remove(capture())
                firstValue.selectionArgs shouldBe arrayOf(expectedModel.requestModel.id)
            }
        }

        verify(coreCompletionHandler).onError(expectedId, expectedModel)
        verify(mockWorker).unlock()
        verify(mockWorker).run()
        verifyNoMoreInteractions(mockWorker)
    }

    @Test
    fun testOnError_4xx_callsHandlerPost() {
        middleware.concurrentHandlerHolder = concurrentHandlerHolder
        val threadSpy = ThreadSpy<Unit>()
        whenever(mockWorker.unlock()).thenAnswer(threadSpy)
        middleware.onError(expectedId, createResponseModel(401))

        threadSpy.verifyCalledOnCoreSdkThread()
        verify(mockWorker, timeout(50)).run()
    }

    @Test
    fun testOnError_408_shouldHandleErrorAsRetriable() {
        val expectedModel = createResponseModel(408)

        middleware.onError(expectedId, expectedModel)

        waitForEventLoopToFinish(concurrentHandlerHolder.coreHandler)
        waitForEventLoopToFinish(uiHandler)

        verify(mockWorker).unlock()
        verifyNoMoreInteractions(mockWorker)
        verifyNoInteractions(coreCompletionHandler)
        verifyNoInteractions(requestRepository)
    }

    @Test
    fun testOnError_429_shouldHandleErrorAsRetriable() {
        val expectedModel = createResponseModel(429)

        middleware.onError(expectedId, expectedModel)

        waitForEventLoopToFinish(concurrentHandlerHolder.coreHandler)
        waitForEventLoopToFinish(uiHandler)

        verify(mockWorker).unlock()
        verifyNoMoreInteractions(mockWorker)
        verifyNoInteractions(coreCompletionHandler)
        verifyNoInteractions(requestRepository)
    }

    @Test
    fun testOnError_5xx() {
        val expectedModel = createResponseModel(500)

        middleware.onError(expectedId, expectedModel)

        waitForEventLoopToFinish(concurrentHandlerHolder.coreHandler)
        waitForEventLoopToFinish(uiHandler)

        verify(mockWorker).unlock()
        verifyNoMoreInteractions(mockWorker)
        verifyNoInteractions(coreCompletionHandler)
        verifyNoInteractions(requestRepository)
    }

    @Test
    fun testOnError_4xx_withCompositeModel() {
        val ids = Array(2040) { i -> "id$i" }
        val responseModel = ResponseModel.Builder()
            .statusCode(400)
            .message("Bad Request")
            .headers(HashMap())
            .body("{'key': 'value'}")
            .requestModel(createRequestModel(ids))
            .build()

        middleware.onError("0", responseModel)

        waitForEventLoopToFinish(concurrentHandlerHolder.coreHandler)
        waitForEventLoopToFinish(uiHandler)

        argumentCaptor<String>().apply {
            verify(middleware.coreCompletionHandler, times(2040))!!.onError(
                capture(),
                eq(responseModel)
            )
            allValues shouldBe ids.toList()
        }
    }

    @Test
    fun testOnError_4xx_withCompositeModel_withManyRequests_whenModulo500Is0() {
        val ids = Array(2000) { i -> "id$i" }

        val responseModel = ResponseModel.Builder()
            .statusCode(400)
            .message("Bad Request")
            .headers(HashMap())
            .body("{'key': 'value'}")
            .requestModel(createRequestModel(ids))
            .build()
        middleware.onError("0", responseModel)

        waitForEventLoopToFinish(concurrentHandlerHolder.coreHandler)
        waitForEventLoopToFinish(uiHandler)

        argumentCaptor<FilterByRequestIds>().apply {
            runBlocking {
                verify(requestRepository, times(40)).remove(capture())
                allValues.size shouldBeLessThanOrEqual 500
            }
        }
    }


    @Test
    fun testOnError_4xx_withCompositeModel_withManyRequests_whenModulo500IsNotZero() {
        val ids = Array(2040) { i -> "id$i" }

        val responseModel = ResponseModel.Builder()
            .statusCode(400)
            .message("Bad Request")
            .headers(HashMap())
            .body("{'key': 'value'}")
            .requestModel(createRequestModel(ids))
            .build()

        middleware.onError("0", responseModel)

        waitForEventLoopToFinish(concurrentHandlerHolder.coreHandler)
        waitForEventLoopToFinish(uiHandler)

        argumentCaptor<FilterByRequestIds>().apply {
            runBlocking {
                verify(requestRepository, times(41)).remove(capture())
                allValues.size shouldBeLessThanOrEqual 500
            }
        }
    }

    @Test
    fun testOnError_withException() {
        val expectedException = Exception("Expected exception")

        middleware.onError(expectedId, expectedException)

        waitForEventLoopToFinish(concurrentHandlerHolder.coreHandler)
        waitForEventLoopToFinish(uiHandler)

        verify(mockWorker).unlock()
        verifyNoMoreInteractions(mockWorker)
        verify(coreCompletionHandler).onError(expectedId, expectedException)
        verifyNoMoreInteractions(coreCompletionHandler)
        verifyNoInteractions(requestRepository)
    }

    private fun createRequestModel(ids: Array<String>): RequestModel {
        return CompositeRequestModel(
            "0",
            "https://emarsys.com",
            RequestMethod.POST,
            null,
            HashMap(),
            100,
            900000,
            ids
        )
    }

    private fun createResponseModel(requestModel: RequestModel): ResponseModel {
        return ResponseModel.Builder()
            .statusCode(200)
            .message("OK")
            .headers(HashMap())
            .body("{'key': 'value'}")
            .requestModel(requestModel)
            .build()
    }

    private fun createResponseModel(statusCode: Int): ResponseModel {
        val requestModel = mock<RequestModel>()
        whenever(requestModel.id).thenReturn(expectedId)
        return ResponseModel.Builder()
            .statusCode(statusCode)
            .body("body")
            .message("message")
            .requestModel(requestModel)
            .build()
    }

    private fun waitForEventLoopToFinish(handler: SdkHandler) {
        val latch = CountDownLatch(1)
        handler.post { latch.countDown() }
        latch.await()
    }

    private fun waitForEventLoopToFinish(handler: Handler) {
        val latch = CountDownLatch(1)
        handler.post { latch.countDown() }
        latch.await()
    }
}